### ISSUE TITLE
OIDC Session cookie is not invalid after logout 

### ISSUE THREAD

**Issue by MollyGu-CN** [OpenLiberty/open-liberty#9909]
**Describe the bug**
This comes from pentest team, on of the session invalid scenario. 
The liberty server configured to use oidc authentication process. After logout, the cookie **WAS_********* can still be reused in another browser.

If there is a stack trace, please include the FULL stack trace (without any `[internal classes]` lines in it). To find the full stack trace, you may need to check in `$WLP_OUTPUT_DIR/messages.log`

**Steps to Reproduce**
1. Login in one browser(e.g firefix), export the cookies using Cookie-Editor addon
2. Open another browser(e.g chrome), import previous exported cookie, can directly access application without ask login again.
3. Logout in first browser successfully.
Following this to logout:
https://www.ibm.com/support/knowledgecenter/SS7K4U_liberty/com.ibm.websphere.wlp.zseries.doc/ae/twlp_logout_endpoint.html

4. Second browser can still continue use the cookies(imported in step 2).

**Expected behavior**
After logout in first browser, should invalid the session cookie in oidc server side, and second browser **shouldn't able to use the old cookies** .

**Diagnostic information:**
 - OpenLiberty Version:    
websphere-liberty:webProfile8
 - Java Version: 
default@apm-auth-oidc-client-deployment-5d65fdc79d-f2nhq:/usr/share/java$ /opt/ibm/java/jre/bin/java -version
java version "1.8.0_221"
Java(TM) SE Runtime Environment (build 8.0.5.41 - pxa6480sr5fp41-20190919_01(SR5 FP41))
IBM J9 VM (build 2.9, JRE 1.8.0 Linux amd64-64-Bit Compressed References 20190911_427071 (JIT enabled, AOT enabled)
OpenJ9   - d581d49
OMR      - ca4db84
IBM      - 5cfdf9c)
JCL - 20190918_01 based on Oracle jdk8u221-b11

 - server.xml configuration (WITHOUT sensitive information like passwords)

OIDC liberty server configuration :
<server description="new server">

    
    <featureManager>
        
        <feature>localConnector-1.0</feature>
        
        <feature>openidConnectServer-1.0</feature>
        <feature>oauth-2.0</feature>
        <feature>localConnector-1.0</feature>
        <feature>ssl-1.0</feature>
        <feature>ldapRegistry-3.0</feature>
        
       
       ${MaximoUserRegistryExtension}
    </featureManager>
    <logging consoleLogLevel="INFO" consoleFormat="basic"/>
    
    <httpSession cookieSecure="true" cookiePath="/vm" idReuse="false"/>
    <webAppSecurity logoutOnHttpSessionExpire="true" trackLoggedOutSSOCookies="true" singleSignonEnabled="true" ssoRequiresSSL="true" preserveFullyQualifiedReferrerUrl="true" ></webAppSecurity>
    
    <webContainer enablePostOnlyJSecurityCheck="true"></webContainer>
    
    
    <httpEndpoint httpPort="9080" httpsPort="9443" id="defaultHttpEndpoint" host="*" />
    
    ${leftBracket_Maximo} baseURL="${MaximoBaseURL}" cacheExpireInSecond="500" admin="${MaximoAdminName}" password="${MaximoAdminPassword}">
    ${rightBracket_Maximo}
    
    ${LdapRegistryContent}
  	${BasicRegistryContent}

    
    <applicationManager autoExpand="true"/>

    <applicationMonitor updateTrigger="mbean"/>

    <oauth-roles>
		<authenticated>
			<special-subject type="ALL_AUTHENTICATED_USERS"/>
		</authenticated>
	</oauth-roles>

    
	<openidConnectProvider id="${oauthProviderId}" issuerIdentifier="${oidcIssuerIdentifier}" oauthProviderRef="${oauthProviderRef}" signatureAlgorithm="${signatureAlgorithm}" keyStoreRef="${keyStoreRef}" keyAliasName="${keyAliasName}" />

	<oauthProvider customLoginURL="${oidcProviderCustomLoginURL}"
	               id="${oauthProviderRef}"
	               jwtAccessToken="true">
		<localStore>
			<client displayname="${openidConnectClientDisplayname}"
			        enabled="true"
			        name="${openidConnectClientName}"
			        preAuthorizedScope="openid profile email phone address"
			        redirect="${oidcClientRedirectURL}"
			        scope="openid"
			        secret="${openidConnectClientSecret}">
			</client>
		</localStore>
	</oauthProvider>
	
    ${leftBracket_Ldap_SSL}
    <ssl id="LDAPSSLSettings" keyStoreRef="ldapKeyStore"/> 
    <keyStore id="ldapKeyStore" location="${ldapKeyStoreLocation}" type="JKS" password="${defaultKeyStorePassword}" /> 
    ${rightBracket_Ldap_SSL}
    
	<sslDefault sslRef="defaultSSLConfig" />

  
	<ssl id="defaultSSLConfig" keyStoreRef="defaultKeyStore" trustStoreRef="defaultTrustStore"  securityLevel="HIGH" clientAuthenticationSupported="true" sslProtocol="TLSv1.2"/>
	
	<keyStore id="defaultKeyStore" password="${defaultKeyStorePassword}"	location="${defaultKeyStoreLocation}" pollingRate="500ms" updateTrigger="polled"/>
	
	<keyStore id="defaultTrustStore" password="${defaultTrustStorePassword}"	location="${defaultTrustStoreLocation}" pollingRate="500ms" updateTrigger="polled"/>
    
  
  
  <webApplication id="apm-oidc-provider" location="apm-oidc-provider.war" contextRoot="apm-oidc-provider/${tenantId}"  name="apm-oidc-provider"/>
</server>

 - If it would be useful, upload the messages.log file found in `$WLP_OUTPUT_DIR/messages.log`

**Additional context**

**Comment by chunlongliang-ibm** [comment]
please add

<webAppSecurity trackLoggedOutSSOCookies="true"/>

to server.xml

**Comment by chunlongliang-ibm** [comment]
are you logout from OP or RP?

**Comment by MollyGu-CN** [comment]
> please add
> 
> <webAppSecurity trackLoggedOutSSOCookies="true"/>
> 
> to server.xml

Already added if you check the server.xml I attached

**Comment by MollyGu-CN** [comment]
> please add
> 
> <webAppSecurity trackLoggedOutSSOCookies="true"/>
> 
> to server.xml

already added, already attached the server.xml sample

**Comment by chunlongliang-ibm** [comment]
logout from OP or RP?

**Comment by MollyGu-CN** [comment]
> are you logout from OP or RP?

Following this link to logout https://www.ibm.com/support/knowledgecenter/SS7K4U_liberty/com.ibm.websphere.wlp.zseries.doc/ae/twlp_logout_endpoint.html

Links looks like below:
OP logout:
https://<hostname>/oidc/endpoint/op-tenant1/logout
RP logout sample code looks like below :
`	@GET
	@Path("/logout/")
	public Response logout() {
		ResponseBuilder rs=Response.ok("Logout successfully");
		try {
			servletRequest.logout();		
			Cookie[] cookies = servletRequest.getCookies();
		    if (cookies != null)
		        for (Cookie cookie : cookies) {
//					System.out.println("Name: "+cookie.getName()+" & Value: "+cookie.getValue());
		            NewCookie newCookie = new NewCookie(cookie.getName(), "", "/", null, 0, null, 0, true);
		            rs.cookie(newCookie);
		        }
	        
		    HttpSession session = servletRequest.getSession();
	        session.invalidate();
     
		} catch (ServletException e) {
			// TODO Auto-generated catch block
			e.printStackTrace();
		}
		return  rs.build();
	}`

**Comment by chunlongliang-ibm** [comment]
can you show me the code where RP redirect user to OP logout?

**Comment by MollyGu-CN** [comment]
> can you show me the code where RP redirect user to OP logout?

It's like I showed the logout endpoint for OP:
`  //oidc server set by vm, need invoke the endpoint : /oidc/endpoint/"+tenantId+"/logout
                        axios.get("/oidc/endpoint/"+tenantId+"/logout").then(() => {});`

**Comment by MollyGu-CN** [comment]
Are you checking on this issue ? @arunavemulapalli

**Comment by arunavemulapalli** [comment]
Investigating

**Comment by ayoho** [comment]
The kind of logic we likely need for correctly deleting cookies can be found in the change I delivered here for #11479: https://github.com/OpenLiberty/open-liberty/pull/11629/files#diff-28943fd61359671e53d4117670b312c3R178

**Comment by jensleis** [comment]
Any updates on this? We are currently seeing the same behavior for one of our apps. Thx

**Comment by ayoho** [comment]
@MollyGu-CN I've opened #19995 to deliver some tests related to this scenario. The test I've added specifically to recreate your scenario appears to be working without any modifications to the product code. That test does the following:
- Logs in through the normal OIDC flow
- Copies the resulting cookies (including SSO cookies from the RP and OP) to a new session
- Logs out in the original session and verifies that the SSO cookies have been removed
- Attempts to access the protected resource in the other session that has the copied SSO cookies

The final step is not granted access to the protected resource and is redirected to log in, as expected. The logout portion of my RP application in my tests essentially just calls:
```Java
request.logout();
response.sendRedirect("https://<host>:<port>/oidc/endpoint/<provider-id>/end_session");
```
